File: libatsc3Module/src/main/java/org/ngbp/libatsc3/ndk/a331/LLSParserSLT.java
package org.ngbp.libatsc3.ndk.a331;

import android.util.Log;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;
import org.xmlpull.v1.XmlPullParserFactory;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.ArrayList;

public class LLSParserSLT {

    public ArrayList<Service> parseXML(String myXml) {
        ArrayList<Service> services = new ArrayList<>();
        XmlPullParserFactory parserFactory;
        try {
            parserFactory = XmlPullParserFactory.newInstance();
            XmlPullParser parser = parserFactory.newPullParser();
            //InputStream is = getAssets().open("data.xml");
            parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false);
            parser.setInput(new ByteArrayInputStream(myXml.getBytes()), null);

            processParsing(services, parser);

        } catch (XmlPullParserException | IOException e) {
            Log.e("LLSParserSLT","exception in parsing: "+e);
        }

        return services;
    }

    private void processParsing(ArrayList<Service> services, XmlPullParser parser) throws IOException, XmlPullParserException {
        Service currentService = null;
        BroadcastSvcSignaling broadcastSvcSignaling = null;
        String attrVal;

        int eventType = parser.getEventType();
        while (eventType != XmlPullParser.END_DOCUMENT) {
            if (eventType == XmlPullParser.START_TAG) {
                String eltName = parser.getName();

                if ("Service".equals(eltName)) {
                    currentService = new Service();
                    services.add(currentService);

                    if ((attrVal = parser.getAttributeValue(null, "serviceId")) != null) {
                        currentService.serviceId = Integer.parseInt(attrVal);
                    }
                    if ((attrVal = parser.getAttributeValue(null, "globalServiceID")) != null) {
                        currentService.globalServiceId = attrVal;
                    }
                    if ((attrVal = parser.getAttributeValue(null, "majorChannelNo")) != null) {
                        currentService.majorChannelNo = Integer.parseInt(attrVal);
                    }
                    if ((attrVal = parser.getAttributeValue(null, "minorChannelNo")) != null) {
                        currentService.minorChannelNo = Integer.parseInt(attrVal);
                    }
                    if ((attrVal = parser.getAttributeValue(null, "shortServiceName")) != null) {
                        currentService.shortServiceName = attrVal;
                    }
                } else if ("BroadcastSvcSignaling".equals(eltName)) {
                    if (currentService != null) {
                        broadcastSvcSignaling = new BroadcastSvcSignaling();
                        currentService.broadcastSvcSignalingCollection.add(broadcastSvcSignaling);

                        if ((attrVal = parser.getAttributeValue(null, "slsProtocol")) != null) {
                            broadcastSvcSignaling.slsProtocol = Integer.parseInt(attrVal);
                        }
                        if ((attrVal = parser.getAttributeValue(null, "slsMajorProtocolVersion")) != null) {
                            broadcastSvcSignaling.slsMajorProtocolVersion = Integer.parseInt(attrVal);
                        }
                        if ((attrVal = parser.getAttributeValue(null, "slsMinorProtocolVersion")) != null) {
                            broadcastSvcSignaling.slsMinorProtocolVersion = Integer.parseInt(attrVal);
                        }
                        if ((attrVal = parser.getAttributeValue(null, "slsDestinationIpAddress")) != null) {
                            broadcastSvcSignaling.slsDestinationIpAddress = attrVal;
                        }
                        if ((attrVal = parser.getAttributeValue(null, "slsDestinationUdpPort")) != null) {
                            broadcastSvcSignaling.slsDestinationUdpPort = attrVal;
                        }
                        if ((attrVal = parser.getAttributeValue(null, "slsSourceIpAddress")) != null) {
                            broadcastSvcSignaling.slsSourceIpAddress = attrVal;
                        }
                    }
                }
            }

            eventType = parser.next();
        }

    }
}

File: libatsc3Module/src/main/java/org/ngbp/libatsc3/entities/service/LLSParserSLT.kt
package org.ngbp.libatsc3.entities.service

import android.util.Log
import org.ngbp.libatsc3.utils.XmlUtils
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.IOException
import java.util.*

class LLSParserSLT {
    fun parseXML(xmlPayload: String): ArrayList<Atsc3Service> {
        val services = ArrayList<Atsc3Service>()
        try {
            val parser = XmlUtils.newParser(xmlPayload)
            parser.nextTag()
            parser.require(XmlPullParser.START_TAG, null, "SLT")

            while (parser.next() != XmlPullParser.END_DOCUMENT) {
                if (parser.eventType != XmlPullParser.START_TAG) {
                    continue
                }
                val name = parser.name
                if (name == ENTRY_SERVICE) {
                    services.add(readService(parser))
                } else {
                    XmlUtils.skip(parser)
                }
            }
        } catch (e: XmlPullParserException) {
            Log.e("LLSParserSLT", "exception in parsing: $e")
        } catch (e: IOException) {
            Log.e("LLSParserSLT", "exception in parsing: $e")
        }
        return services
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun readService(parser: XmlPullParser): Atsc3Service {
        val currentService = Atsc3Service()

        val attrCount = parser.attributeCount
        for (i in 0 until attrCount) {
            when (parser.getAttributeName(i)) {
                "serviceId" -> currentService.serviceId = XmlUtils.strToInt(parser.getAttributeValue(i))
                "globalServiceID" -> currentService.globalServiceId = parser.getAttributeValue(i)
                "majorChannelNo" -> currentService.majorChannelNo = XmlUtils.strToInt(parser.getAttributeValue(i))
                "minorChannelNo" -> currentService.minorChannelNo = XmlUtils.strToInt(parser.getAttributeValue(i))
                "shortServiceName" -> currentService.shortServiceName = parser.getAttributeValue(i)
                else -> {
                    // skip attribute
                }
            }
        }

        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }

            val name = parser.name
            if (name == "BroadcastSvcSignaling") {
                currentService.broadcastSvcSignalingCollection.add(readBroadcastSvcSignaling(parser))
            } else {
                XmlUtils.skip(parser)
            }
        }

        return currentService
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun readBroadcastSvcSignaling(parser: XmlPullParser): BroadcastSvcSignaling {
        val broadcastSvcSignaling = BroadcastSvcSignaling()

        val attrCount = parser.attributeCount
        for (i in 0 until attrCount) {
            val value = parser.getAttributeValue(i)
            when (parser.getAttributeName(i)) {
                "slsProtocol" -> broadcastSvcSignaling.slsProtocol = XmlUtils.strToInt(value)
                "slsMajorProtocolVersion" -> broadcastSvcSignaling.slsMajorProtocolVersion = XmlUtils.strToInt(value)
                "slsMinorProtocolVersion" -> broadcastSvcSignaling.slsMinorProtocolVersion = XmlUtils.strToInt(value)
                "slsDestinationIpAddress" -> broadcastSvcSignaling.slsDestinationIpAddress = value
                "slsDestinationUdpPort" -> broadcastSvcSignaling.slsDestinationUdpPort = value
                "slsSourceIpAddress" -> broadcastSvcSignaling.slsSourceIpAddress = value
                else -> {
                    // skip attribute
                }
            }
        }

        return broadcastSvcSignaling
    }

    companion object {
        private const val ENTRY_SERVICE = "Service"
    }
}